dsp/src/lockin.rs

```diff
@@ -0,0 +1,544 @@
+        let mut in_phase = [0f32; ADC_SAMPLE_BUFFER_SIZE];
+        let mut quadrature = [0f32; ADC_SAMPLE_BUFFER_SIZE];
+        in_phase
+            .iter_mut()
+            .zip(quadrature.iter_mut())
+            .zip(adc_samples.iter())
+            .enumerate()
+            .for_each(|(n, ((i, q), sample))| {
+                let integer_phase: i32 = (n as i32 * self.sample_period as i32
+                    - self.timestamps[0].unwrap())
+                    * self.harmonic as i32;
+                let phase = self.phase_offset
+                    + 2. * PI * integer_phase as f32 / reference_period as f32;
+                let (sine, cosine) = libm::sincosf(phase);
+                let sample = *sample as f32;
+                *i = sine * sample;
+                *q = cosine * sample;
+            });
+
+        Ok((in_phase, quadrature))
+    }
+
+    /// Filter the in-phase and quadrature signals using the supplied
+    /// biquad IIR. The signal arrays are modified in place.
+    ///
+    /// # Arguments
+    ///
+    /// * `in_phase` - In-phase signal.
+    /// * `quadrature` - Quadrature signal.
+    pub fn filter(&mut self, in_phase: &mut [f32], quadrature: &mut [f32]) {
+        in_phase
+            .iter_mut()
+            .zip(quadrature.iter_mut())
+            .for_each(|(i, q)| {
+                *i = self.iir[0].update(&mut self.iirstate[0], *i);
+                *q = self.iir[1].update(&mut self.iirstate[1], *q);
+            });
+    }
+}
+
+/// Decimate the in-phase and quadrature signals to
+/// `DECIMATED_BUFFER_SIZE`. The ratio of `ADC_SAMPLE_BUFFER_SIZE` to
+/// `DECIMATED_BUFFER_SIZE` must be a power of 2.
+///
+/// # Arguments
+///
+/// * `in_phase` - In-phase signal.
+/// * `quadrature` - Quadrature signal.
+///
+/// # Returns
+///
+/// The decimated in-phase and quadrature signals.
+pub fn decimate(
+    in_phase: [f32; ADC_SAMPLE_BUFFER_SIZE],
+    quadrature: [f32; ADC_SAMPLE_BUFFER_SIZE],
+) -> ([f32; DECIMATED_BUFFER_SIZE], [f32; DECIMATED_BUFFER_SIZE]) {
+    let n_k = ADC_SAMPLE_BUFFER_SIZE / DECIMATED_BUFFER_SIZE;
+    debug_assert!(
+        ADC_SAMPLE_BUFFER_SIZE == DECIMATED_BUFFER_SIZE || n_k % 2 == 0
+    );
+
+    let mut in_phase_decimated = [0f32; DECIMATED_BUFFER_SIZE];
+    let mut quadrature_decimated = [0f32; DECIMATED_BUFFER_SIZE];
+
+    in_phase_decimated
+        .iter_mut()
+        .zip(quadrature_decimated.iter_mut())
+        .zip(
+            in_phase
+                .iter()
+                .step_by(n_k)
+                .zip(quadrature.iter().step_by(n_k)),
+        )
+        .for_each(|((i_d, q_d), (i, q))| {
+            *i_d = *i;
+            *q_d = *q;
+        });
+
+    (in_phase_decimated, quadrature_decimated)
+}
+
+/// Compute the magnitude and phase from the in-phase and quadrature
+/// signals. The in-phase and quadrature arrays are modified in place.
+///
+/// # Arguments
+///
+/// * `in_phase` - In-phase signal.
+/// * `quadrature` - Quadrature signal.
```

Comment:
It's not clear where magnitude and phase are stored. Which magnitude and which phase?

I think this would be simpler if you took ownership fo the `in_phase` and `quadrature` arrays and then returned them with the new values as (magnitude, phase) arrays. It would still be in-place updates, but ownership would make it clear that the data has transformed.

E.g.
```rust
let (i, q) = //...;
let (mag, phase) = dsp::lockin::magnitude_phase(i, q);
```

Reply:
The difficulty with this is that this could be performed on arrays of various sizes. For instance, `magnitude_phase` might be called after `decimate`. Alternatively `decimate` might be omitted.

Reply:
Based on Robert's feedback, it looks like this will go away anyway.